unmarshaller.go

```diff
@@ -1,117 +1,114 @@
-package gocsv
+package commando
 
 import (
-	"encoding/csv"
 	"fmt"
+	"io"
 	"reflect"
 )
 
 // Unmarshaller is a CSV to struct unmarshaller.
 type Unmarshaller struct {
-	reader                 *csv.Reader
-	Headers                []string
-	fieldInfoMap           []*fieldInfo
-	MismatchedHeaders      []string
-	MismatchedStructFields []string
-	outType                reflect.Type
+	config *validConfig
+	line   int
+	reader Reader
 }
 
-// NewUnmarshaller creates an unmarshaller from a csv.Reader and a struct.
-func NewUnmarshaller(reader *csv.Reader, out interface{}) (*Unmarshaller, error) {
+// NewUnmarshaller is a convenience function which allocates and
+// returns a new Unmarshaller.
+func NewUnmarshaller(holder interface{}, reader Reader) (*Unmarshaller, error) {
+	return (&Config{Holder: holder}).NewUnmarshaller(reader)
+}
+
+// NewUnmarshaller creates an unmarshaller from a Reader and a struct.
+func (c *Config) NewUnmarshaller(reader Reader) (*Unmarshaller, error) {
 	headers, err := reader.Read()
 	if err != nil {
 		return nil, err
 	}
-	headers = normalizeHeaders(headers)
 
-	um := &Unmarshaller{reader: reader, outType: reflect.TypeOf(out)}
-	err = validate(um, out, headers)
+	vc, err := c.validate(headers)
 	if err != nil {
 		return nil, err
 	}
+
+	um := &Unmarshaller{
+		reader: reader,
+		config: vc,
+		line:   1,
+	}
+
 	return um, nil
 }
 
-// Read returns an interface{} whose runtime type is the same as the struct that
-// was used to create the Unmarshaller.
+// Read returns an interface{} whose runtime type is the same as the
+// struct that was used to create the Unmarshaller.
 func (um *Unmarshaller) Read() (interface{}, error) {
-	row, err := um.reader.Read()
-	if err != nil {
-		return nil, err
+	value, _, err := um.ReadUnmatched()
+	return value, err
+}
+
+func ReadAllCallback(um *Unmarshaller, cb func(interface{}) error) error {
+	for {
+		rec, err := um.Read()
+		if err == io.EOF {
+			return nil
+		} else if err != nil {
+			return err
+		}
```

Comment:
Having mixed feelings about this as well.  A definite goal is to let the caller decide what to do about errors, which this … doesn't.  The easy answer is to have the callback be `func(interface{}, error) error` and have it return `error` if it wants to stop processing.

I don't love this, since it means the straightforward two-line append-and-return-nil type callbacks get bloated with boilerplatey error checking/logging.

Maybe this should have two callbacks, ala:

```go
func ReadAllCallback(um *Unmarshaller, onSuccess func(interface{}) error, onError func(error) error) error
``` 

I also don't love this, but it does have the nice property of shoving all the error handling stuff into reusable functions, ala:

```go
func AbortOnAnyError(e error) error {
    return e
}

func LogAndConitnue(e error) error {
    log.Error(e, "Failed to unmarshal")
    return nil
}
```

Reply:
I don't think it makes sense to make store the error callback in the unmarshaller/config if you are passing the success callback.  I would think both should be passed in each time. 

I am still not clear on what kind of errors OnError would actually be handling?

Reply:
> I don't think it makes sense to make store the error callback in the unmarshaller/config if you are passing the success callback. I would think both should be passed in each time.

Reasonable, but it sure is fiddly.  So calls with this approach would look like:

```go
	func LogAndContinue(err error) error {
		log.Error(ctx, "Failed to unmarshal record", err, nil)
		return nil
	}

	func StopOnError(err error) error {
		return err
	}

	err = commando.ReadAllCallback(um,
		func(rec interface{}) error {
			records = append(records, rec.(ConcreteType))
			return nil
		}, LogAndContinue,
	)

	err = commando.ReadAllCallback(um,
		func(rec interface{}) error {
			records = append(records, rec.(ConcreteType))
			return nil
		}, StopOnError,
	)
```

`StopOnError` could be part of Commando, the log function could be put in our `lib/` somewhere, since it depends on our internal logging code.

> I am still not clear on what kind of errors OnError would actually be handling?

Failures to unmarshal a row of CSV into a struct, for example, because we expect a float, but the value is `"$100" ` or `50-` or some other invalid string.  This type of error handling would let us parse all the valid rows within a file, while skipping invalid ones — whereas we currently fail the entire file if a single record cannot be unmarshalled.

Reply:
I understand.  What you have here doesn't look so terrible to me.